ghcide/src/Text/Fuzzy/Parallel.hs

```diff
@@ -1,96 +1,90 @@
+match :: T.Text    -- ^ Pattern in lowercase except for first character
+      -> T.Text    -- ^ The text to search in.
+      -> Maybe Int -- ^ The score
+match (T.Text pArr pOff pLen) (T.Text sArr sOff sLen) = go 0 1 pOff sOff
   where
-    null :: (T.TextualMonoid s) => s -> Bool
-    null = not . T.any (const True)
-
-    s = extract t
-    (totalScore, _currScore, result, pat, _) =
-      T.foldl'
-        undefined
-        (\(tot, cur, res, pat, isFirst) c ->
-            case T.splitCharacterPrefix pat of
-              Nothing -> (tot, 0, res <> T.singleton c, pat, isFirst)
-              Just (x, xs) ->
-                -- the case of the first character has to match
-                -- otherwise use lower case since the pattern is assumed lower
-                let !c' = if isFirst then c else toLower c in
-                if x == c' then
-                  let cur' = cur * 2 + 1 in
-                  (tot + cur', cur', res <> pre <> T.singleton c <> post, xs, False)
-                else (tot, 0, res <> T.singleton c, pat, isFirst)
-        ) ( 0
-          , 1 -- matching at the start gives a bonus (cur = 1)
-          , mempty, pattern, True) s
+    pTotal = pOff + pLen
+    sDelta = sOff + sLen - pTotal
+
+    go !totalScore !currScore !currPOff !currSOff
+      -- If pattern has been matched in full
+      | currPOff >= pTotal
+      = Just totalScore
+      -- If there is not enough left to match the rest of the pattern, equivalent to
+      -- (sOff + sLen - currSOff) < (pOff + pLen - currPOff)
+      | currSOff > currPOff + sDelta
+      = Nothing
+      -- This is slightly broken for non-ASCII:
+      -- 1. If code units, consisting a single pattern code point, are found as parts
+      --    of different code points, it counts as a match. Unless you use a ton of emojis
+      --    as identifiers, such false positives should not be be a big deal,
+      --    and anyways HLS does not currently support such use cases, because it uses
+      --    code point and UTF-16 code unit positions interchangeably.
+      -- 2. Case conversions is not applied to non-ASCII code points, because one has
+      --    to call T.toLower (not T.map toLower), reallocating the string in full, which
+      --    is too much of performance penalty for fuzzy search. Again, anyway HLS does not
+      --    attempt to do justice to Unicode: proper Unicode text matching requires
+      --    `unicode-transforms` and friends.
+      -- Altogether we sacrifice correctness for the sake of performance, which
+      -- is a right trade-off for fuzzy search.
+      | pByte <- TA.unsafeIndex pArr currPOff
+      , sByte <- TA.unsafeIndex sArr currSOff
+      -- First byte (currPOff == pOff) should match exactly, otherwise - up to case.
+      , pByte == sByte || (currPOff /= pOff && pByte == toLowerAscii sByte)
+      = let curr = currScore * 2 + 1 in
+        go (totalScore + curr) curr (currPOff + 1) (currSOff + 1)
+      | otherwise
```

Comment:
Well, even the old implementation was not fully matching `fuzzy` package because of a different approach to case matching. What's the best place to add unit tests?

Reply:
The ghcide test suite.